Gradle for Java plugin breaks with Gradle 9 projects

Extension Name: vscode-gradle
Extension Version: 3.16.4
OS Version: Windows 11 24H2 (Windows_NT x64 10.0.26100)
VSCode version: 1.102.3

**Describe the bug**
The Gradle for Java plugin breaks with Gradle 9 projects. When loading the project in VS Code, the Gradle for Java plugin shows "Gradle: Build Error" at the bottom and the output tab shows errors.

I think this is the most relevant bit, but I might be wrong:
```
[error] FAILURE: Build failed with an exception.

* What went wrong:
org.gradle.api.plugins.internal.DefaultDecoratedConvention
```

**To Reproduce**
1. Create a new project with `gradle init`, using Gradle 9.0.0
2. In VS Code, open the folder containing the new project
3. **BUG**: The initial project import fails with a build error, expected Gradle tasks (e.g. build, test, etc) are not available in the sidebar.

**Expected behavior**
The Gradle for Java plugin works with Gradle 9 projects.

**Screenshots**
n/a

**Output from "Gradle for Java"**
```
[info] [gradle-server] Gradle Server started, listening on 50123
[info] Gradle client connected to server
[info] Java Home: C:\Users\user2233\.vscode\extensions\redhat.java-1.44.0-win32-x64\jre\21.0.8-win32-x86_64
[info] JVM Args: -XX:MaxMetaspaceSize=384m,-XX:+HeapDumpOnOutOfMemoryError,-Xms256m,-Xmx512m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant
[info] Gradle User Home: C:\Users\user2233\.gradle
[info] Gradle Version: 9.0.0
[error] FAILURE: Build failed with an exception.

* What went wrong:
org.gradle.api.plugins.internal.DefaultDecoratedConvention
> org.gradle.api.plugins.internal.DefaultDecoratedConvention

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

CONFIGURE FAILED in 23s
[error] [gradle-server] The supplied build action failed with an exception.
[error] Error getting build for c:\Users\user2233\dev\gradle-vscode-test: The supplied build action failed with an exception.
[info] Found 0 tasks
[info] Java Home: C:\Users\user2233\.vscode\extensions\redhat.java-1.44.0-win32-x64\jre\21.0.8-win32-x86_64
[info] JVM Args: -XX:MaxMetaspaceSize=384m,-XX:+HeapDumpOnOutOfMemoryError,-Xms256m,-Xmx512m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant
[info] Gradle User Home: C:\Users\user2233\.gradle
[info] Gradle Version: 9.0.0
[error] FAILURE: Build failed with an exception.

* What went wrong:
org.gradle.api.plugins.internal.DefaultDecoratedConvention
> org.gradle.api.plugins.internal.DefaultDecoratedConvention

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

CONFIGURE FAILED in 770ms
[error] [gradle-server] The supplied build action failed with an exception.
[error] Error getting build for c:\Users\user2233\dev\gradle-vscode-test: The supplied build action failed with an exception.
```

**Does the bug still exist if you disable all other extensions?**
Yes

**Additional context**

Gradle 9.0.0 was just released on 2025-07-31: https://docs.gradle.org/9.0.0/release-notes.html. So this might be more of a feature request for Gradle 9 support, rather than a bug. Thank you to the maintainers for this plugin!

--- a/build.gradle
+++ b/build.gradle
@@ -8,7 +8,7 @@ plugins {
 project.ext.set('grpcVersion', '1.53.0')
 project.ext.set('protobufVersion', '3.12.0')
 project.ext.set('protocVersion', project.protobufVersion)
-project.ext.set('toolingAPIVersion', '8.8')
+project.ext.set('toolingAPIVersion', '9.0.0')
 
 allprojects {
   group = 'vscode-gradle'

--- a/gradle-plugin/src/main/java/com/microsoft/gradle/GradleProjectModelBuilder.java
+++ b/gradle-plugin/src/main/java/com/microsoft/gradle/GradleProjectModelBuilder.java
@@ -34,9 +34,6 @@ import org.gradle.api.artifacts.result.ResolvedDependencyResult;
 import org.gradle.api.initialization.dsl.ScriptHandler;
 import org.gradle.api.internal.initialization.DefaultScriptHandler;
 import org.gradle.api.internal.tasks.TaskContainerInternal;
-import org.gradle.api.plugins.Convention;
-import org.gradle.api.plugins.ExtensionsSchema;
-import org.gradle.api.plugins.ExtensionsSchema.ExtensionSchema;
 import org.gradle.api.reflect.TypeOf;
 import org.gradle.api.tasks.JavaExec;
 import org.gradle.api.tasks.TaskContainer;
@@ -177,8 +174,13 @@ public class GradleProjectModelBuilder implements ToolingModelBuilder {
 	}
 
 	private List<String> getPlugins(Project project) {
-		Convention convention = project.getConvention();
-		return new ArrayList<>(convention.getPlugins().keySet());
+		// The Convention API was removed in Gradle 8.0+
+		// Use the PluginContainer instead to get applied plugins
+		List<String> plugins = new ArrayList<>();
+		project.getPlugins().forEach(plugin -> {
+			plugins.add(plugin.getClass().getName());
+		});
+		return plugins;
 	}
 
 	private List<GradleClosure> getPluginClosures(Project project) {
@@ -186,36 +188,41 @@ public class GradleProjectModelBuilder implements ToolingModelBuilder {
 				.compareTo(GradleVersion.version(MINIMAL_SUPPORTED_PLUGIN_CLOSURE_VERSION)) < 0) {
 			return Collections.emptyList();
 		}
-		Convention convention = project.getConvention();
-		ExtensionsSchema extensionsSchema = convention.getExtensionsSchema();
+		// The Convention API was removed in Gradle 8.0+
+		// Use the ExtensionContainer instead to get extension schemas
 		List<GradleClosure> closures = new ArrayList<>();
-		for (ExtensionSchema schema : extensionsSchema.getElements()) {
-			TypeOf<?> publicType = schema.getPublicType();
-			Class<?> concreteClass = publicType.getConcreteClass();
-			List<GradleMethod> methods = new ArrayList<>();
-			List<GradleField> fields = new ArrayList<>();
-			for (Method method : concreteClass.getMethods()) {
-				String name = method.getName();
-				List<String> parameterTypes = new ArrayList<>();
-				for (Class<?> parameterType : method.getParameterTypes()) {
-					parameterTypes.add(parameterType.getName());
+		try {
+			project.getExtensions().getExtensionsSchema().getElements().forEach(schema -> {
+				TypeOf<?> publicType = schema.getPublicType();
+				Class<?> concreteClass = publicType.getConcreteClass();
+				List<GradleMethod> methods = new ArrayList<>();
+				List<GradleField> fields = new ArrayList<>();
+				for (Method method : concreteClass.getMethods()) {
+					String name = method.getName();
+					List<String> parameterTypes = new ArrayList<>();
+					for (Class<?> parameterType : method.getParameterTypes()) {
+						parameterTypes.add(parameterType.getName());
+					}
+					methods.add(new DefaultGradleMethod(name, parameterTypes, isDeprecated(method)));
+					int modifiers = method.getModifiers();
+					// See:
+					// https://docs.gradle.org/current/userguide/custom_gradle_types.html#managed_properties
+					// we offer managed properties for an abstract getter method
+					if (name.startsWith("get") && name.length() > 3 && Modifier.isPublic(modifiers)
+							&& Modifier.isAbstract(modifiers)) {
+						fields.add(new DefaultGradleField(name.substring(3, 4).toLowerCase() + name.substring(4),
+								isDeprecated(method)));
+					}
 				}
-				methods.add(new DefaultGradleMethod(name, parameterTypes, isDeprecated(method)));
-				int modifiers = method.getModifiers();
-				// See:
-				// https://docs.gradle.org/current/userguide/custom_gradle_types.html#managed_properties
-				// we offer managed properties for an abstract getter method
-				if (name.startsWith("get") && name.length() > 3 && Modifier.isPublic(modifiers)
-						&& Modifier.isAbstract(modifiers)) {
-					fields.add(new DefaultGradleField(name.substring(3, 4).toLowerCase() + name.substring(4),
-							isDeprecated(method)));
+				for (Field field : concreteClass.getFields()) {
+					fields.add(new DefaultGradleField(field.getName(), isDeprecated(field)));
 				}
-			}
-			for (Field field : concreteClass.getFields()) {
-				fields.add(new DefaultGradleField(field.getName(), isDeprecated(field)));
-			}
-			DefaultGradleClosure closure = new DefaultGradleClosure(schema.getName(), methods, fields);
-			closures.add(closure);
+				DefaultGradleClosure closure = new DefaultGradleClosure(schema.getName(), methods, fields);
+				closures.add(closure);
+			});
+		} catch (Exception e) {
+			// Fallback to empty list if extensions schema is not available
+			return Collections.emptyList();
 		}
 		return closures;
 	}

